### PR TITLE
Prepare placeholders before timing out

### PR DESCRIPTION
**Pull request by acdlite**
While a tree is suspended, prepare for the timeout by pre-rendering the placeholder state.

This simplifies the implementation a bit because every render now results in a completed tree.

### CODE REVIEW

**Comment by react-linter** [comment]
ReactDOM: size: :small_red_triangle:+0.5%, gzip: :small_red_triangle:+0.4%


  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: 97af3e1f3a6130dc38269b8389da057152868be2...dc6b87c5fcf5209c17d037dbf9e2bd1bd79e81e8</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | +0.4% | +0.4% | 626.27 KB | 629.04 KB | 146.33 KB | 146.98 KB | UMD_DEV
**react-dom.production.min.js** | **:small_red_triangle:+0.5%** | **:small_red_triangle:+0.4%** | **95.33 KB** | **95.78 KB** | **30.82 KB** | **30.94 KB** | **UMD_PROD**
react-dom.development.js | +0.4% | +0.5% | 622.4 KB | 625.17 KB | 145.15 KB | 145.81 KB | NODE_DEV
**react-dom.production.min.js** | **:small_red_triangle:+0.5%** | **:small_red_triangle:+0.4%** | **95.32 KB** | **95.77 KB** | **30.35 KB** | **30.48 KB** | **NODE_PROD**
ReactDOM-dev.js | +0.5% | +0.5% | 629.62 KB | 632.62 KB | 143.78 KB | 144.46 KB | FB_WWW_DEV
**ReactDOM-prod.js** | **:small_red_triangle:+0.7%** | **:small_red_triangle:+0.5%** | **272.35 KB** | **274.12 KB** | **51.29 KB** | **51.55 KB** | **FB_WWW_PROD**
react-dom.profiling.min.js | +0.5% | +0.4% | 96.22 KB | 96.67 KB | 30.65 KB | 30.79 KB | NODE_PROFILING

## react-art
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-art.development.js | +0.8% | +0.9% | 410.55 KB | 413.94 KB | 91.94 KB | 92.76 KB | UMD_DEV
**react-art.production.min.js** | **:small_red_triangle:+0.6%** | **:small_red_triangle:+0.7%** | **82.54 KB** | **83.05 KB** | **25.41 KB** | **25.58 KB** | **UMD_PROD**
react-art.development.js | +1.0% | +1.1% | 342.53 KB | 345.91 KB | 74.81 KB | 75.63 KB | NODE_DEV
**react-art.production.min.js** | **:small_red_triangle:+1.1%** | **:small_red_triangle:+1.1%** | **47.52 KB** | **48.03 KB** | **14.8 KB** | **14.97 KB** | **NODE_PROD**
ReactART-dev.js | +1.1% | +1.1% | 332.56 KB | 336.2 KB | 69.89 KB | 70.68 KB | FB_WWW_DEV
**ReactART-prod.js** | **:small_red_triangle:+1.4%** | **:small_red_triangle:+1.0%** | **143.43 KB** | **145.45 KB** | **24.65 KB** | **24.9 KB** | **FB_WWW_PROD**

## react-test-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-test-renderer.development.js | +1.0% | +1.1% | 344.8 KB | 348.18 KB | 74.75 KB | 75.58 KB | UMD_DEV
**react-test-renderer.production.min.js** | **:small_red_triangle:+1.1%** | **:small_red_triangle:+1.0%** | **47.44 KB** | **47.95 KB** | **14.55 KB** | **14.69 KB** | **UMD_PROD**
react-test-renderer.development.js | +1.0% | +1.1% | 340.92 KB | 344.3 KB | 73.79 KB | 74.61 KB | NODE_DEV
**react-test-renderer.production.min.js** | **:small_red_triangle:+1.1%** | **:small_red_triangle:+1.3%** | **47.14 KB** | **47.65 KB** | **14.37 KB** | **14.55 KB** | **NODE_PROD**
ReactTestRenderer-dev.js | +1.1% | +1.1% | 344.08 KB | 347.71 KB | 72.54 KB | 73.36 KB | FB_WWW_DEV

## react-reconciler
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-reconciler.development.js | +0.8% | +1.0% | 333.05 KB | 335.82 KB | 71.28 KB | 71.96 KB | NODE_DEV
**react-reconciler.production.min.js** | **:small_red_triangle:+1.0%** | **:small_red_triangle:+1.0%** | **47.24 KB** | **47.7 KB** | **14.17 KB** | **14.3 KB** | **NODE_PROD**
react-reconciler-persistent.development.js | +0.8% | +1.0% | 331.66 KB | 334.44 KB | 70.71 KB | 71.39 KB | NODE_DEV
**react-reconciler-persistent.production.min.js** | **:small_red_triangle:+1.0%** | **:small_red_triangle:+1.0%** | **47.25 KB** | **47.71 KB** | **14.17 KB** | **14.31 KB** | **NODE_PROD**

## react-native-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
ReactNativeRenderer-dev.js | +0.6% | +0.6% | 466.76 KB | 469.7 KB | 102.7 KB | 103.34 KB | RN_FB_DEV
**ReactNativeRenderer-prod.js** | **:small_red_triangle:+0.9%** | **:small_red_triangle:+0.7%** | **207.42 KB** | **209.33 KB** | **36.53 KB** | **36.8 KB** | **RN_FB_PROD**
ReactNativeRenderer-dev.js | +0.6% | +0.6% | 466.47 KB | 469.41 KB | 102.64 KB | 103.28 KB | RN_OSS_DEV
**ReactNativeRenderer-prod.js** | **:small_red_triangle:+0.5%** | **:small_red_triangle:+0.9%** | **199.98 KB** | **200.99 KB** | **35.06 KB** | **35.36 KB** | **RN_OSS_PROD**
ReactFabric-dev.js | +0.6% | +0.6% | 457.01 KB | 459.96 KB | 100.3 KB | 100.95 KB | RN_FB_DEV
**ReactFabric-prod.js** | **:small_red_triangle:+0.5%** | **:small_red_triangle:+0.9%** | **192.24 KB** | **193.25 KB** | **33.64 KB** | **33.93 KB** | **RN_FB_PROD**
ReactFabric-dev.js | +0.6% | +0.6% | 457.04 KB | 459.99 KB | 100.32 KB | 100.96 KB | RN_OSS_DEV
**ReactFabric-prod.js** | **:small_red_triangle:+0.5%** | **:small_red_triangle:+0.9%** | **192.28 KB** | **193.29 KB** | **33.66 KB** | **33.94 KB** | **RN_OSS_PROD**
ReactNativeRenderer-profiling.js | +0.5% | +0.8% | 202.52 KB | 203.53 KB | 35.61 KB | 35.9 KB | RN_OSS_PROFILING
ReactFabric-profiling.js | +0.5% | +0.8% | 194.49 KB | 195.51 KB | 34.13 KB | 34.41 KB | RN_OSS_PROFILING

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by mzangie** [comment]
Hm

**Comment by blling** [comment]
I see some big PRs related to `placeholder`.
What is `placeholder` in react? 
Could you give me any docs link about react `placeholder`?

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 1]
```diff
@@ -785,34 +786,7 @@ function propagateContextChange<V>(
         if (fiber.type === context && (observedBits & changedBits) !== 0) {
           // Update the expiration time of all the ancestors, including
           // the alternates.
-          let node = fiber;
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: Uninline because this is a really hot path

File: packages/react-reconciler/src/ReactFiberRoot.js [hunk 2]
```diff
@@ -95,6 +95,7 @@ export function createFiberRoot(
 
     pendingCommitExpirationTime: NoWork,
     finishedWork: null,
+    timeoutID: -1,
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberRoot.js]: TODO: Add `emptyTimeoutHandle`, `setTimeout`, and `clearTimeout` to HostConfig

File: packages/react-reconciler/src/ReactFiberScheduler.js [hunk 3]
```diff
@@ -1132,7 +1105,114 @@ function renderRoot(root: FiberRoot, isYieldy: boolean): void {
...
+  interruptedBy = null;
+
+  if (nextRenderDidError) {
+    // There was an error
+    if (hasLowerPriorityWork(root, expirationTime)) {
+      // There's lower priority work. If so, it may have the effect of fixing
+      // the exception that was just thrown. Exit without committing. This is
+      // similar to a suspend, but without a timeout because we're not waiting
+      // for a promise to resolve. React will restart at the lower
+      // priority level.
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberScheduler.js]: TODO: Hoist comment so it's clear that it applies to the `else if` block as well.

File: packages/react-reconciler/src/ReactFiberScheduler.js [hunk 4]
```diff
@@ -1299,89 +1348,144 @@ function markTimeout(
...
-    requestWork(root, retryTime);
+function retrySuspendedRoot(
+  root: FiberRoot,
+  fiber: Fiber,
+  suspendedTime: ExpirationTime,
+) {
+  // A ping is different from an update. We don't need to traverse the entire
+  // ancestor path, because we already have access to the root. We can bail out
+  // once we reach a parent with sufficient priority.
+  scheduleWorkToNearestMatchingAncestor(fiber, suspendedTime);
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberScheduler.js]: TODO: Use the same function as `scheduleWork`

File: packages/react-reconciler/src/ReactFiberScheduler.js [hunk 5]
```diff
@@ -1905,10 +2044,13 @@ function completeRoot(
 // When working on async work, the reconciler asks the renderer if it should
 // yield execution. For DOM, we implement this with requestIdleCallback.
 function shouldYield() {
-  if (deadline === null || deadlineDidExpire) {
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberScheduler.js]: TODO: Move this to separate PR and add unit test that would have caught this

File: packages/react-reconciler/src/ReactFiberUnwindWork.js [hunk 6]
```diff
@@ -160,113 +147,83 @@ function throwException(
...
-    if (elapsedMs < 0) {
-      elapsedMs = 0;
-    }
-    const remainingTimeMs = expirationTimeMs - currentTimeMs;
-
     // Find the earliest timeout of all the timeouts in the ancestor path.
-    // TODO: Alternatively, we could store the earliest timeout on the context
-    // stack, rather than searching on every suspend.
+    // TODO: We should store the earliest timeout on the context stack, rather
+    // than traversing the return path on every suspend.
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberUnwindWork.js]: TODO: Don't need to store this on the stack. We only hit this path if we're IO-bound. Using the stack would slow down even the non-IO-bound case.

File: packages/react-reconciler/src/ReactFiberScheduler.js [hunk 7]
```diff
@@ -1844,12 +1958,22 @@ function performWorkOnRoot(
     }
   } else {
     // Flush async work.
+    const isYieldy = true;
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberScheduler.js]: TODO: Move next to the caller

File: packages/react-reconciler/src/ReactFiberScheduler.js [hunk 8]
```diff
@@ -1828,14 +1928,28 @@ function performWorkOnRoot(
   isRendering = true;
 
   // Check if this is async work or sync/expired work.
-  if (!isYieldy) {
+  if (deadline === null || isExpired) {
     // Flush work without yielding.
+    // TODO: Non-yieldy work does not necessarily imply expired work. A renderer
+    // may want to perform some work without yielding, but also without
+    // requiring the root to complete (by triggering placeholders).
+    const isYieldy = false;
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberScheduler.js]: TODO: Move next to the caller
